ydb/library/yql/providers/yt/provider/yql_yt_provider_context.cpp

```diff
@@ -0,0 +1,214 @@
+    EJoinKind /*joinKind*/) {
+    if (joinAlgo == EJoinAlgoType::LookupJoin) {
+        return !leftJoinKeys.empty() && IsLookupJoinApplicable(*left->Stats, *right->Stats, leftJoinKeys);
+    }
+    if (joinAlgo == EJoinAlgoType::MapJoin) {
+        return IsMapJoinApplicable(*right->Stats);
+    }
+    return joinAlgo == EJoinAlgoType::MergeJoin;
+}
+
+TOptimizerStatistics TYtProviderContext::ComputeJoinStats(
+    const TOptimizerStatistics& leftStats,
+    const TOptimizerStatistics& rightStats,
+    const TVector<TString>& leftJoinKeys,
+    const TVector<TString>& rightJoinKeys,
+    EJoinAlgoType joinAlgo,
+    EJoinKind /*joinKind*/,
+    TCardinalityHints::TCardinalityHint* /*maybeHint*/) const {
+
+    const TYtProviderStatistic* leftSpecific = static_cast<const TYtProviderStatistic*>(leftStats.Specific.get());
+    const TYtProviderStatistic* rightSpecific = static_cast<const TYtProviderStatistic*>(rightStats.Specific.get());
+
+    TMaybe<double> maxUniques;
+    auto leftMaxUniques = FindMaxUniqueVals(*leftSpecific, leftJoinKeys);
+    if (leftMaxUniques && (!maxUniques || *maxUniques < *leftMaxUniques)) {
+        maxUniques = *leftMaxUniques;
+    }
+    auto rightMaxUniques = FindMaxUniqueVals(*rightSpecific, rightJoinKeys);
+    if (rightMaxUniques && (!maxUniques || *maxUniques < *rightMaxUniques)) {
+        maxUniques = *rightMaxUniques;
+    }
+
+    auto resultSpecific = std::make_unique<TYtProviderStatistic>();
+
+    resultSpecific->RelBitmap = leftSpecific->RelBitmap | rightSpecific->RelBitmap;
+    resultSpecific->JoinAlgo = joinAlgo;
+
+    double outputCardinality = ComputeCardinality(maxUniques, leftStats.Nrows, rightStats.Nrows);
+    double leftCardinalityFactor = leftStats.Nrows != 0.0 ? outputCardinality / leftStats.Nrows : 0.0;
+    double rightCardinalityFactor = rightStats.Nrows != 0.0 ? outputCardinality / rightStats.Nrows : 0.0;
+
+    double outputByteSize = leftCardinalityFactor * leftStats.ByteSize + rightCardinalityFactor * rightStats.ByteSize;
+
+    auto leftReadBytes = leftStats.ByteSize;
+    double rightReadBytes = rightStats.ByteSize;
+
+    if (joinAlgo == EJoinAlgoType::LookupJoin) {
+        auto leftJoinUniques = FindMaxUniqueVals(*leftSpecific, TVector<TString>{leftJoinKeys[0]});
+        auto rightJoinUniques = FindMaxUniqueVals(*rightSpecific, TVector<TString>{rightJoinKeys[0]});
+        if (IsLookupJoinApplicable(leftStats, rightStats, leftJoinKeys) &&
+            leftJoinUniques && rightJoinUniques && *rightJoinUniques < *leftJoinUniques) {
+            leftReadBytes *= *rightJoinUniques / *leftJoinUniques;
+        }
+    }
+
+    if (joinAlgo == EJoinAlgoType::MapJoin &&
+        (leftSpecific->JoinAlgo == EJoinAlgoType::MapJoin || leftSpecific->JoinAlgo == EJoinAlgoType::LookupJoin) &&
+        rightStats.Type == EStatisticsType::BaseTable) {
+        // Optimistically assume that consecutive MapJoins are fused together.
+        leftReadBytes = 0;
+    }
+
+    double outputCost =
```

Comment:
So this might be ok for the first iteration, but really you need separate cost functions for different join algorithms

Reply:
Ack
Right now the cost function is simple enough now to include all join algorithms.